// eslint-plugin-vuejs-accessibility/dist/rules/mouse-events-have-key-events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const rule = {
    meta: {
        docs: {
            url: utils_1.makeDocsURL("mouse-events-have-key-events")
        },
        messages: {
            mouseOver: "@mouseover, @mouseenter, or @hover must be accompanied by @focusin or @focus for accessibility.",
            mouseOut: "@mouseout or @mouseleave must be accompanied by @focusout or @blur for accessibility."
        }
    },
    create(context) {
        return utils_1.defineTemplateBodyVisitor(context, {
            VElement(node) {
                if (utils_1.hasOnDirectives(node, ["mouseover", "mouseenter", "hover"]) &&
                    !utils_1.hasOnDirectives(node, ["focus", "focusin"])) {
                    context.report({ node: node, messageId: "mouseOver" });
                }
                if (utils_1.hasOnDirectives(node, ["mouseout", "mouseleave"]) &&
                    !utils_1.hasOnDirectives(node, ["blur", "focusout"])) {
                    context.report({ node: node, messageId: "mouseOut" });
                }
            }
        });
    }
};
exports.default = rule;
